thunder/core/update_aliases.py

```diff
@@ -0,0 +1,107 @@
+
+
+def _involves_viewed_args(bsym, viewed):
+    if bsym.sym.id == prims.PrimIDs.RETURN:
+        return False
+    return any(isinstance(p, TensorProxy) and variableify(p) in viewed for p in bsym.flat_proxy_args)
+
+
+def insert_alias_updates(computation_trace: Trace) -> list[Trace]:
+    swap_map = dict()
+    bsyms = []
+
+    # First pass: identify views, their originals, and operands involved in inplace ops
+    view_groups = []
+    inplace_inputs = set()
+    for bsym in computation_trace.bound_symbols:
+        if _is_inplace_op(bsym) or _is_view_creation_op(bsym):
+            # only interested in the input which is modified by the inplace op
+            in_tensor = list(map(variableify, filter(lambda p: isinstance(p, TensorProxy), bsym.flat_proxy_args)))[0]
+            out_tensors = set(map(variableify, filter(lambda p: isinstance(p, TensorProxy), bsym.flat_proxy_outs)))
+            if _is_inplace_op(bsym):
+                inplace_inputs.add(in_tensor)
+            for group in view_groups:
+                if in_tensor in group:
+                    group.update(out_tensors)
+                    group.add(in_tensor)
+                    break
+            else:
+                view_groups.append(out_tensors.union({in_tensor}))
+
+    # filter out view groups that don't have any tensors involved in inplace ops
+    view_groups = [group for group in view_groups if len(group.intersection(inplace_inputs)) != 0]
+    viewed = set(reduce(set.union, view_groups, set()))
+    encountered = set()
+    for bsym in computation_trace.bound_symbols:
+        if _is_inplace_op(bsym) or _is_view_creation_op(bsym) or _involves_viewed_args(bsym, viewed):
+            in_tensors = map(variableify, filter(lambda p: isinstance(p, TensorProxy), bsym.flat_proxy_args))
+            if _is_inplace_op(bsym):
+                in_tensors = {list(in_tensors)[0]}
+            else:
+                in_tensors = set(in_tensors)
+            out_tensors = set(map(variableify, filter(lambda p: isinstance(p, TensorProxy), bsym.flat_proxy_outs)))
+            encountered.update(in_tensors)
+            group = next((g for g in view_groups if any(g.intersection(in_tensors))), None)
+            if group is None:
+                # this is a view creation with operands that are not involved in any inplace ops
+                bsyms.append(bsym.from_bsym_swap_proxies(swap_map))
+                continue
+            views_encountered = group.intersection(encountered)
+            new_aliases = _get_new_aliases(views_encountered, computation_trace)
+
+            update_bsym, swap_map = _get_update_bsym(views_encountered, swap_map, new_aliases)
+            bsyms.append(update_bsym)
+            encountered.update(out_tensors)
+            new_bsym = bsym.from_bsym_swap_proxies(swap_map)
+            if _is_inplace_op(bsym):
+                #  This relies on these being one element sets
+                swap_map[in_tensors.pop()] = unvariableify(out_tensors.pop())
```

Comment:
I'm reading your changes from top, so far I'm not quote following the point of having one `in_tensor` as a set of one element. Would there be some (future) use cases or extensions of it?

Reply:
In the case that the bsym has views as arguments, in_tensors may have more than one element.  For example:
```
def f(x, y):
    a = x.view(-1)
    b = a.mul(y)
    return b
```
Then the bsym for mul would have a and y as in_tensors, and because a is a view, it will pass through the update_alias machine.  But in the line you reference, the assumption is that the bsym is an in-place op and so its input and its output should each have only one element.